src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -45,27 +44,119 @@ static ucs_status_t uct_cuda_ipc_mkey_pack(uct_md_h md, uct_mem_h memh,
+    int64_t *b = (int64_t *) dst->bytes;
+    *b++ = *a++;
+    *b   = *a;
+}
+
+static ucs_status_t uct_cuda_ipc_get_unique_index_for_uuid(int* idx,
+                                                           uct_cuda_ipc_md_t* md,
+                                                           CUuuid* uuid)
+{
+    int i;
+
+    for (i = 0; i < md->uuid_map_size; i++) {
+        if (uct_cuda_ipc_uuid_equals(uuid, &md->uuid_map[i])) {
+            *idx = i;
+            return UCS_OK; /* found */
+        }
+    }
+
+    if (ucs_unlikely(md->uuid_map_size == md->uuid_map_capacity)) {
+        /* reallocate on demand */
+        int num_devices;
+        int original_cache_size, new_cache_size;
+        int new_capacity = md->uuid_map_capacity * 2;
+
+        UCT_CUDA_IPC_DEVICE_GET_COUNT(num_devices);
+        original_cache_size   = md->uuid_map_capacity * num_devices;
+        new_cache_size        = new_capacity * num_devices;
+        md->uuid_map_capacity = new_capacity;
+        md->uuid_map          = ucs_realloc(md->uuid_map,
+                                            new_capacity * sizeof(CUuuid),
+                                            "uct_cuda_ipc_uuid_map");
+        if (md->uuid_map == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        md->peer_accessible_cache = ucs_realloc(md->peer_accessible_cache,
+                                                new_cache_size,
+                                                "uct_cuda_ipc_peer_accessible_cache");
+        if (md->peer_accessible_cache == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        memset(md->peer_accessible_cache + original_cache_size, 0xFF,
+               new_cache_size - original_cache_size);
+    }
+
+    /* Add new mapping */
+    uct_cuda_ipc_uuid_copy(&md->uuid_map[md->uuid_map_size], uuid);
+    *idx = md->uuid_map_size++;
 
     return UCS_OK;
 }
 
+static ucs_status_t uct_cuda_ipc_is_peer_accessible(uct_cuda_ipc_component_t *mdc,
+                                                    uct_cuda_ipc_key_t *rkey)
+{
+    CUdevice this_device;
+    ucs_status_t status;
+    int peer_idx;
+    int num_devices;
+    char* accessible;
+
+    status = uct_cuda_ipc_get_unique_index_for_uuid(&peer_idx, mdc->md, &rkey->uuid);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    UCT_CUDA_IPC_GET_DEVICE(this_device);
+    UCT_CUDA_IPC_DEVICE_GET_COUNT(num_devices);
+
+    accessible = &mdc->md->peer_accessible_cache[peer_idx * num_devices + this_device];
+    if (*accessible == -1) { /* unchecked, add to cache */
+        /* rkey->d_mapped is picked up in uct_cuda_ipc_cache_map_memhandle */
+        CUresult result = cuIpcOpenMemHandle(&rkey->d_mapped,
+                                             rkey->ph,
+                                             CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS);
+        *accessible = (result != CUDA_SUCCESS && result != CUDA_ERROR_ALREADY_MAPPED)
```

Comment:
plz, use `()` for each statement: `((result != CUDA_SUCCESS) && (result != CUDA_ERROR_ALREADY_MAPPED))`